### ISSUE TITLE
aci_access_port_to_interface_policy_leaf_profile reports changed when child infraPortBlk exist

### ISSUE THREAD

**Issue by mtorelli** [ansible/ansible#43106]
##### SUMMARY

When using the aci_access_port_to_interface_policy_leaf_profile module to create two separate port blocks (infraPortBlk) under the same Access Port Selector (infraHPortS) and both port blocks already exists, one of them will report status "ok" and the other one status "changed".  Both should report status "ok" since both already exist.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

aci_access_port_to_interface_policy_leaf_profile
##### ANSIBLE VERSION

```
ansible 2.6.1
  config file = None
  configured module search path = [u'/u/mtorelli/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /u/mtorelli/gitlab/gdl2_refresh/venv_2_6/lib/python2.7/site-packages/ansible
  executable location = /u/mtorelli/gitlab/gdl2_refresh/venv_2_6/bin/ansible
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### OS / ENVIRONMENT

Red Hat Enterprise Linux Server release 7.5 (Maipo)
Cisco APIC version 3.1(2m)
##### STEPS TO REPRODUCE

Run the two tasks below to create the two port blocks under the access port selector and then re-run the same tasks.  On the second run, one of the tasks will show status "ok" and the other one "changed"

```yaml
- name: Adding port 1/34 to access port selector MyPortSelector
  aci_access_port_to_interface_policy_leaf_profile:
    hostname: "{{ apic_hostname }}"
    username: "{{ apic_username }}"
    password: "{{ apic_password }}"
    use_ssl: yes
    validate_certs: no
    access_port_selector: MyPortSelector
    from: 34
    to: 34
    leaf_interface_profile: MyLeafProfile
    policy_group: MyPolicyGroup
    interface_type: port_channel
    leaf_port_blk: 30534
    description: Access Port Selector for policy-group MyPolicyGroup
    leaf_port_blk_description: MyDescription
    state: present

- name: Adding port 1/33 to access port selector MyPortSelector
  aci_access_port_to_interface_policy_leaf_profile:
    hostname: "{{ apic_hostname }}"
    username: "{{ apic_username }}"
    password: "{{ apic_password }}"
    use_ssl: yes
    validate_certs: no
    access_port_selector: MyPortSelector
    from: 33
    to: 33
    leaf_interface_profile: MyLeafProfile
    policy_group: MyPolicyGroup
    interface_type: port_channel
    leaf_port_blk: 30533
    description: Access Port Selector for policy-group MyPolicyGroup
    leaf_port_blk_description: MyDescription
    state: present

```

##### EXPECTED RESULTS

On first run, both tasks should report "changed" since neither object exists.   Both tasks should report status "ok" on all subsequent runs. 
##### ACTUAL RESULTS

```
(venv_2_6) aci-tools:ansible:% ansible-playbook -i inventory wichita.yml -vv
ansible-playbook 2.6.1
  config file = None
  configured module search path = [u'/u/mtorelli/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /u/mtorelli/gitlab/gdl2_refresh/venv_2_6/lib/python2.7/site-packages/ansible
  executable location = /u/mtorelli/gitlab/gdl2_refresh/venv_2_6/bin/ansible-playbook
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
No config file found; using defaults
statically imported: /u/mtorelli/gitlab/gdl2_refresh/ansible/roles/pc_interfaces/tasks/add_port_select_tasks.yml

PLAYBOOK: wichita.yml *********************************************************************************************************************
1 plays in wichita.yml

PLAY [Configuring ICT Staging APIC] *******************************************************************************************************
META: ran handlers

TASK [pc_interfaces : Adding port 1/34 to access port selector MyPortSelector] ************************************************************
task path: /u/mtorelli/gitlab/gdl2_refresh/ansible/roles/pc_interfaces/tasks/add_port_select_tasks.yml:3
changed: [ict_apic_dev] => {"changed": true, "current": [{"infraHPortS": {"attributes": {"descr": "Access Port Selector for policy-group MyPolicyGroup", "dn": "uni/infra/accportprof-MyLeafProfile/hports-MyPortSelector-typ-range", "name": "MyPortSelector", "nameAlias": "", "ownerKey": "", "ownerTag": "", "type": "range"}, "children": [{"infraPortBlk": {"attributes": {"descr": "MyDescription33", "fromCard": "1", "fromPort": "33", "name": "30533", "nameAlias": "", "toCard": "1", "toPort": "33"}}}, {"infraPortBlk": {"attributes": {"descr": "MyDescription34", "fromCard": "1", "fromPort": "34", "name": "30534", "nameAlias": "", "toCard": "1", "toPort": "34"}}}, {"infraRsAccBaseGrp": {"attributes": {"fexId": "101", "tDn": "uni/infra/funcprof/accbundle-MyPolicyGroup"}}}]}}]}

TASK [pc_interfaces : Adding port 1/33 to access port selector MyPortSelector] ************************************************************
task path: /u/mtorelli/gitlab/gdl2_refresh/ansible/roles/pc_interfaces/tasks/add_port_select_tasks.yml:21
ok: [ict_apic_dev] => {"changed": false, "current": [{"infraHPortS": {"attributes": {"descr": "Access Port Selector for policy-group MyPolicyGroup", "dn": "uni/infra/accportprof-MyLeafProfile/hports-MyPortSelector-typ-range", "name": "MyPortSelector", "nameAlias": "", "ownerKey": "", "ownerTag": "", "type": "range"}, "children": [{"infraPortBlk": {"attributes": {"descr": "MyDescription33", "fromCard": "1", "fromPort": "33", "name": "30533", "nameAlias": "", "toCard": "1", "toPort": "33"}}}, {"infraPortBlk": {"attributes": {"descr": "MyDescription34", "fromCard": "1", "fromPort": "34", "name": "30534", "nameAlias": "", "toCard": "1", "toPort": "34"}}}, {"infraRsAccBaseGrp": {"attributes": {"fexId": "101", "tDn": "uni/infra/funcprof/accbundle-MyPolicyGroup"}}}]}}]}```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/network/aci/aci_access_port_to_interface_policy_leaf_profile.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/aci/aci_access_port_to_interface_policy_leaf_profile.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @brunocalogero @dagwieers @jmcgill298 @schunduri
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
cc @rsmeyers

**Comment by smnmtzgr** [comment]
I notice the same problem.

**Comment by dagwieers** [comment]
In this case the value for the `access_port_selector` parameter needs to be unique, because that's the internal identifier. If you use the same in both, they will basically overwrite the same object.

If [the documentation](https://docs.ansible.com/ansible/devel/modules/aci_access_port_to_interface_policy_leaf_profile_module.html) is not clear on this, please propose changes to the documentation by editing the online documentation (see **Edit on GitHub** link on top).

Please let me know if this fixes your problem.

need_info

**Comment by mtorelli** [comment]
Yes, what you suggest is correct.  If you use a single access_port_selector and then add only one port-block to it and then repeat that Ansible task, it will report "ok" rather than "changed".   However, it is a legitimate configuration to use a single selector and continue to add distinct port-blocks to it.  You can see from the verbose output below that the current state matches the desired state which means that it should report "ok".  The code only seems to recognize the first infraPortBlk in the list as the current state.  In other words, no matter how many separate port-blocks you have in the same interface selector it reports "ok" when you try to reapply the first one in the list and changed for all others.  In the example below, since port 36 shows up first in the list it reports "ok"  when you try to reapply it, but ports 33, 34 and 35 all report "change" when you try to reapply them.
```
ok: [ict_apic_dev] => {
    "changed": false, 
    "current": [
        {
            "infraHPortS": {
                "attributes": {
                    "descr": "Access Port Selector for policy-group MyPolicyGroup", 
                    "dn": "uni/infra/accportprof-MyLeafProfile/hports-MyPortSelector-typ-range", 
                    "name": "MyPortSelector", 
                    "nameAlias": "", 
                    "ownerKey": "", 
                    "ownerTag": "", 
                    "type": "range"
                }, 
                "children": [
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription36", 
                                "fromCard": "1", 
                                "fromPort": "36", 
                                "name": "30536", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "36"
                            }
                        }
                    }, 
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription35", 
                                "fromCard": "1", 
                                "fromPort": "35", 
                                "name": "30535", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "35"
                            }
                        }
                    }, 
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription33", 
                                "fromCard": "1", 
                                "fromPort": "33", 
                                "name": "30533", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "33"
                            }
                        }
                    }, 
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription34", 
                                "fromCard": "1", 
                                "fromPort": "34", 
                                "name": "30534", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "34"
                            }
                        }
                    }, 
                    {
                        "infraRsAccBaseGrp": {
                            "attributes": {
                                "fexId": "101", 
                                "tDn": "uni/infra/funcprof/accbundle-MyPolicyGroup"
                            }
                        }
                    }
                ]
            }
        }
    ], 
    "invocation": {
        "module_args": {
            "access_port_selector": "MyPortSelector", 
            "certificate_name": null, 
            "description": "Access Port Selector for policy-group MyPolicyGroup", 
            "from": 36, 
            "from_card": null, 
            "from_port": "36", 
            "host": "foo", 
            "hostname": "foo", 
            "interface_type": "port_channel", 
            "leaf_interface_profile": "MyLeafProfile", 
            "leaf_port_blk": "30536", 
            "leaf_port_blk_description": "MyDescription36", 
            "output_level": "normal", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "policy_group": "MyPolicyGroup", 
            "port": null, 
            "private_key": null, 
            "protocol": "https", 
            "state": "present", 
            "timeout": 30, 
            "to": 36, 
            "to_card": null, 
            "to_port": "36", 
            "use_proxy": true, 
            "use_ssl": true, 
            "username": "ansible", 
            "validate_certs": false
        }
    }
}

changed: [ict_apic_dev] => {
    "changed": true, 
    "current": [
        {
            "infraHPortS": {
                "attributes": {
                    "descr": "Access Port Selector for policy-group MyPolicyGroup", 
                    "dn": "uni/infra/accportprof-MyLeafProfile/hports-MyPortSelector-typ-range", 
                    "name": "MyPortSelector", 
                    "nameAlias": "", 
                    "ownerKey": "", 
                    "ownerTag": "", 
                    "type": "range"
                }, 
                "children": [
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription36", 
                                "fromCard": "1", 
                                "fromPort": "36", 
                                "name": "30536", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "36"
                            }
                        }
                    }, 
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription35", 
                                "fromCard": "1", 
                                "fromPort": "35", 
                                "name": "30535", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "35"
                            }
                        }
                    }, 
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription33", 
                                "fromCard": "1", 
                                "fromPort": "33", 
                                "name": "30533", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "33"
                            }
                        }
                    }, 
                    {
                        "infraPortBlk": {
                            "attributes": {
                                "descr": "MyDescription34", 
                                "fromCard": "1", 
                                "fromPort": "34", 
                                "name": "30534", 
                                "nameAlias": "", 
                                "toCard": "1", 
                                "toPort": "34"
                            }
                        }
                    }, 
                    {
                        "infraRsAccBaseGrp": {
                            "attributes": {
                                "fexId": "101", 
                                "tDn": "uni/infra/funcprof/accbundle-MyPolicyGroup"
                            }
                        }
                    }
                ]
            }
        }
    ], 
    "invocation": {
        "module_args": {
            "access_port_selector": "MyPortSelector", 
            "certificate_name": null, 
            "description": "Access Port Selector for policy-group MyPolicyGroup", 
            "from": 35, 
            "from_card": null, 
            "from_port": "35", 
            "host": "foo", 
            "hostname": "foo", 
            "interface_type": "port_channel", 
            "leaf_interface_profile": "MyLeafProfile", 
            "leaf_port_blk": "30535", 
            "leaf_port_blk_description": "MyDescription35", 
            "output_level": "normal", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "policy_group": "MyPolicyGroup", 
            "port": null, 
            "private_key": null, 
            "protocol": "https", 
            "state": "present", 
            "timeout": 30, 
            "to": 35, 
            "to_card": null, 
            "to_port": "35", 
            "use_proxy": true, 
            "use_ssl": true, 
            "username": "ansible", 
            "validate_certs": false
        }
    }
}
```

**Comment by dagwieers** [comment]
Ok, I see what you mean. In that case the module should be split into one to create the access port selector, and one to manage the ports assigned to an access port selector. We cannot make one module to manage both parents and children as absent/present doesn't indicate what it applies to.

@rsmeyers @nilloBE Can you provide advice how you want to see this work ? I guess this was incorrectly assessed originally.

**Comment by smnmtzgr** [comment]
@dagwieers I think to separate these functions into two modules would be the best way to solve this issue. Another solution would be to change the way "ok" or "changed" status will be found out. Here the information of @mtorelli are very helpful. All ports (port-blocks) should be considered.

**Comment by mtorelli** [comment]
You're suggesting breaking them up into something like the two json objects below?  One that creates access port selectors and associates a policy-group and a separate one that just creates port blocks inside of a given access port selector.

```json
{
    "infraHPortS": {
        "attributes": {
            "dn": "uni/infra/accportprof-{{LeafProfile}}/hports-{{AccessPortSelector}}-typ-range",
            "name": "{{AccessPortSelector}}",
            "rn": "hports-{{AccessPortSelector}}-typ-range",
            "status": "created,modified"
        },
        "children": [
            {
                "infraRsAccBaseGrp": {
                    "attributes": {
                        "status": "created,modified",
                        "tDn": "uni/infra/funcprof/accbundle-{{PolicyGroup}}"
                    },
                    "children": []
                }
            }
        ]
    }
}
```
```json
{
    "infraPortBlk": {
        "attributes": {
            "dn": "uni/infra/accportprof-{{LeafProfile}}/hports-{{AccessPortSelector}}-typ-range/portblk-{{block_id}}",
           "fromPort": "{{from_port}}",
            "rn": "portblk-{{block_id}}",
            "status": "created,modified",
            "toPort": "{{to_port}}"
        },
        "children": []
    }
}
```

**Comment by smnmtzgr** [comment]
yep right. I think something like this could be a good way to solve this!?

**Comment by rsmeyers** [comment]
Yeah, most use-case we have 1-on-1 but if we wanna have more blocks into 1 access port selector, we need to make an extra module to manage the blocks and find a way to stay backwards compatible

**Comment by smnmtzgr** [comment]
@dagwieers @rsmeyers How can we proceed with this issue? It would be good to get this implemented. If you need some help or information please inform me.

**Comment by dagwieers** [comment]
@smnmtzgr There was no time to refactor this before the v2.7 deadline, so this is scheduled for the v2.8 release. The only thing I need to know to complete this, is the name for this new module. I always leave that up to people with a better understanding and taste of ACI :-)

**Comment by mtorelli** [comment]
Maybe something like "aci_access_port_block_to_access_port_selector" since it's adding/associating the child infraPortBlk to the parent infraHPortS.  Or just "aci_access_port_block" with the parent association implied.

**Comment by smnmtzgr** [comment]
For me "aci_access_port_block_to_access_port_selector" would be the better choice.

**Comment by rsmeyers** [comment]
Let's name it aci_access_port_block_to_access_port to be consistent with the 'parent' module

**Comment by smnmtzgr** [comment]
@dagwieers - Can I try to write the new module and also reduce the functionality of the existing module? If yes, where can I contact you directly to talk about more details?

**Comment by dagwieers** [comment]
@smnmtzgr Of course, that would be great. We need to make sure that the existing module still works as before (but that it gives a DeprecationWarning when this functionality is being used). We can then remove that functionality after 4 new major releases. (So that would be earliest v2.12).

You can reach me either by mail in private (see GitHub account) or on IRC using nick "dag" (but I may not be online all the time).

**Comment by smnmtzgr** [comment]
@dagwieers I just wanted to start writing the new module and tested the existing module of the devel branch of ansible and the issue seems to be resolved without using an additional module. The existing module "aci_access_port_to_interface_policy_leaf_profile" now seems to work correctly, also when doing the things discussed in this issue. The module now seems to report the correct results. Are there already some changes done which could fix this? Eventually not directly in the module itself, but in classes/functions getting used?

 @mtorelli can you also check this? (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/aci/aci_access_port_to_interface_policy_leaf_profile.py)

**Comment by dagwieers** [comment]
@smnmtzgr I wouldn't be surprised if the recent refactoring of the construct_url() logic fixed this too. We discovered a few issues with the old implementation after testing for inconsistencies between 2 codebases.

But that does not take away the need to manage child-objects independently. So the new module is still needed.

**Comment by dagwieers** [comment]
@smnmtzgr I wouldn't be surprised if the recent refactoring of the construct_url() logic (#43441) fixed this too. We discovered a few issues with the old implementation after testing for inconsistencies between 2 codebases.

But that does not take away the need to manage child-objects independently. So the new module is still needed.

**Update:** Bad news. It does not appear to be fixed.

**Comment by smnmtzgr** [comment]
@dagwieers why do you think it is not fixed? Yesterday I tried running my ansible-playbook with ansible 2.6 where I notice the bug/wrong result output. And I also tried running it with ansible devel branch. There everything is working fine. Result is always "ok" for all ports of a port-block, if it already exists. (not only the first one is "ok" -> thats what the bug is all about)

**Comment by dagwieers** [comment]
I used the original example, and saw the original issue again. A subsequent run shows the second task as changed, always.

**Comment by smnmtzgr** [comment]
Hm thats strange... for me the subsequent run shows "ok" for the second task. We need to chat in IRC...

**Comment by dagwieers** [comment]
@smnmtzgr You can try me anytime (but I can't promise to be online all the time though).

**Comment by dagwieers** [comment]
So after debugging together with @smnmtzgr it appears that if we remove `leaf_port_blk_description` the module works correctly on v3.0(2k) and v3.2(2l). However with `leaf_port_blk_description` set, it fails on v3.2(2l), and it reports changed on v3.0(2k), *only* on the subsequent task.

This appears to be a bug in the APIC. Both the difference in behaviour as well as the fact it fails to work.

**Comment by smnmtzgr** [comment]
@dagwieers and me have invested some time for further troubleshooting. 
We found a API bug and we found out that with ansible-devel and also deleting the parameter `leaf_port_blk_description` the issue is fixed.

But nevertheless there are 2 action items now: 
1. Find out more about the parameter `leaf_port_blk_description` and if this is a bug in the API.
2. Write the new ansible module aci_access_port_block_to_access_port, and deprecate the existing module so that it can be changed in the future to not adding/changing the things the new module does. I will write the new module+integration tests and make a PR to ansible-devel.

Just to clear it up:
The additional module is still needed so that we can also safely delete things. We dont have 1:1 relationship between a port selector and an access port block. Its 1:x !
